### ISSUE TITLE
gw2 webform data model

### ISSUE THREAD

**Issue by rburghol** [HARPgroup/vahydro#584]
- webform_submitted_data has all data attached to submittal
- file_managed has pointers to the files uploaded with a webform submittal
- webform_component has entry to each file that is attached to a webform, webform components are keyed by their unique "cid"
  - Example, GW2 cid =
    - gw2 scanned pdf = 136
    - drillers log = 128
    - water quality upload = 123
- Find files for GW-2 webform submittal with sid = 66598:
```
select b.sid, c.name, a.uri,
  str_replace(a.uri, 'public://', '/var/www/html/d.dh/sites/default/files/webform/') as path
from file_managed as a 
left outer join webform_submitted_data as b
on (
  b.data::integer = a.fid
)
left outer join webform_component as c 
on (
  c.cid = b.cid
)
where b.sid = 66598
and c.cid in (123, 128, 136);
```

**Comment by rburghol** [comment]
From Tim:
```
SELECT * -- DISTINCT [wc.name](http://wc.name/)
FROM file_managed fm
LEFT JOIN field_data_field_dha_submittal_document sd on
    fm.fid = sd.field_dha_submittal_document_fid
LEFT JOIN dh_adminreg_feature arf_sub ON
    sd.entity_id = arf_sub.adminid
    AND sd.entity_type = 'dh_adminreg_feature'
LEFT JOIN webform_submitted_data wfsd ON
    wfsd.data::integer = fm.fid
LEFT JOIN webform_component wc ON
    wc.cid = wfsd.cid
WHERE
    wc.cid in (123, 128, 136)
--  AND arf_sub.bundle is null;
```

**Comment by rburghol** [comment]
Variant to find out what file types thaere are:
```
select wc.cid, wc.name, count(*) from file_managed fm
LEFT JOIN field_data_field_dha_submittal_document sd on
    fm.fid = sd.field_dha_submittal_document_fid
LEFT JOIN dh_adminreg_feature arf_sub ON
    sd.entity_id = arf_sub.adminid
    AND sd.entity_type = 'dh_adminreg_feature'
LEFT JOIN webform_submitted_data wfsd ON
    wfsd.data = fm.fid::varchar
LEFT JOIN webform_component wc ON
    wc.cid = wfsd.cid
WHERE wc.type = 'file'
group by wc.cid, wc.name;

```

**Comment by tt-cova-deq** [comment]
@rburghol 

I'm getting hung up on the query above. It doesn't reference the field_data_field_dha_submittal_document table or the dh_adminreg_feature tables. So this query doesn't connect the file_managed record to a submittal in dh_adminreg_feature. If you comment out the first two joins, the query results are exactly the same.

```
select wc.cid, wc.name, count(*) from file_managed fm
LEFT JOIN field_data_field_dha_submittal_document sd on
    fm.fid = sd.field_dha_submittal_document_fid
LEFT JOIN dh_adminreg_feature arf_sub ON
    sd.entity_id = arf_sub.adminid
    AND sd.entity_type = 'dh_adminreg_feature'
LEFT JOIN webform_submitted_data wfsd ON
    wfsd.data = fm.fid::varchar
LEFT JOIN webform_component wc ON
    wc.cid = wfsd.cid
WHERE wc.type = 'file'
group by wc.cid, wc.name;
```

If you add a SELECT for a field in dh_adminreg_feature to the query above (like ftype), there are records with null values for the dh_adminreg_feature, implying the attachment doesn't connect to a submittal. So, I guess my question is how do the Water Quality Results Attachments (attachments with null values for the submittal) specifically relate to the GW2 submittals.

Long story short, the field_data_field_dha_submittal_document seems to connect an attachment to a submittal sometimes but not all of the time. Is there another way that attachments can be associated with a submittal?

**Comment by rburghol** [comment]
Ahh, it looks like the way to link the uploaded doc to a well (is that your goal?), is to find the GW-2 submittal, which points to the GW-2 Webform Submission, which THEN points to uploaded document. 

So this:
> However, the way I'm currently retrieving attachments from a GW2 is dh_feature (well) --> dh_adminreg_feature (submittal) by way of field_data_dh_link_feature_submittal --> file_managed (attachment) by way of field_data_field_dha_submittal_document.

Should Look like this:
> dh_feature (well) --> dh_adminreg_feature (submittal) by way of field_data_dh_link_feature_submittal --> webform_submission by way of dh_link_webform_submittal (webform) --> webform_submitted_data --> file_managed.

So, if the well of interest feature hydroid = 476063, we do this:

```
select well.name, well.hydroid, ar_sub.adminid, 
  wfsd.cid, wfsd.data, replace(fm.uri, 'public://', '/var/www/html/d.dh/sites/default/files/') as path
from dh_feature as well
left outer join field_data_dh_link_feature_submittal sub_link
on (
  sub_link.dh_link_feature_submittal_target_id = well.hydroid
)
left outer join dh_adminreg_feature as ar_sub
on (
  ar_sub.adminid = sub_link.entity_id
)
left outer join field_data_dh_link_webform_submittal as wf_link
on (
  wf_link.entity_id = ar_sub.adminid
)
left outer join webform_submissions as wfs 
on (
  wfs.sid = wf_link.dh_link_webform_submittal_target_id
)
LEFT outer JOIN webform_submitted_data wfsd 
ON (
    wfs.sid = wfsd.sid
)
left outer join file_managed as fm 
on (
  wfsd.cid in (123, 128, 136) 
  and wfsd.data = fm.fid::varchar
)
where hydroid = 476063;

```

Side note: The previous approach we took assumed that the file linkage creation that enables the dh_adminreg_feature to have a direct link to uploaded documents from the webform (instead of going through the webform submission), was used for the uploading of a scanned GW-2 document -- at least that is the only one that is populated in 99.9% of the cases.

I would guess that the population of the field_data_field_dha_submittal_document for anything *other* than the scanned/uploaded paper GW-2 was something that was done later due to a modification request by staff, and something only done for select things like the abandonment form.  Regardless, the robust pathway is that above.   BTW - reminder that  the original business need had the GW-2 as an inviolate snapshot in time, hence these link paths being the robust method -- my guess is that the new architecture does *not* do that, which is something to make note of when passing data to VDH, that is, the data can be potentially volatile.

**Comment by tt-cova-deq** [comment]
@rburghol AHA! Thank you, and thanks for the patience in my getting back to you - busy busy week. 

I just added a join from webform_submitted_data to webform_component to grab the webform component name and I'm getting `Upload GW-2 Form`, `Water Quality Results Attachment`, `Pump Data Information`, and `Upload Drillers Log`! 

This looks really promising. I'll take a deep dive into it tomorrow and make sure it covers everything I'm expecting. Thanks so much!

**Comment by tt-cova-deq** [comment]
And subsequently, this is the query that you would use to get all webform submitted data from Drupal (which appears to only be `gw2_form` and `gw2_abandonment`)?

``` SQL
SELECT DISTINCT
    wfs.sid,
    ar_sub.adminid as adminreg_feature_submission_adminid,
    ar_sub.ftype as adminreg_feature_submission_ftype,
    wc.form_key as webform_component_formkey,
    wc.name as webform_component_name,
    wfsd.data as webform_submission_data
FROM dh_adminreg_feature ar_sub
INNER JOIN field_data_dh_link_webform_submittal wf_link ON
    wf_link.entity_id = ar_sub.adminid
LEFT JOIN webform_submissions AS wfs ON
    wfs.sid = wf_link.dh_link_webform_submittal_target_id
LEFT JOIN webform_submitted_data wfsd ON
    wfs.sid = wfsd.sid
left join webform_component wc ON
	wfsd.cid = wc.cid
    and wfsd.nid = wc.nid
order by ar_sub.adminid
;
```
